### PR TITLE
test: add unit test for user service

### PR DESCRIPTION
**Pull request by takanome-dev**
## Description
This PR adds unit tests for the user service. The unit tests ensure the functionality of the user service is properly tested and validated.

_Generated using [OpenSauced](https://opensauced.ai/)._

## What type of PR is this? (check all applicable)

- [ ] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [x] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: src/user/services/user/user.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        is_open_sauced_member: true,
+        login: supabaseUser.user_metadata.user_name,
+        email: supabaseUser.user_metadata.email,
+        created_at: new Date(supabaseUser.identities[0].created_at),
+        updated_at: new Date(supabaseUser.identities[0].updated_at),
+        connected_at: new Date(supabaseUser.confirmed_at),
+      };
+
+      expect(dbUserRepositoryMock.create).toHaveBeenCalledWith(newUser);
+      // ! need help here, I don't know how to test the `save` method in the `checkAddUser` method (newUser.save()) 🙃
```
Comment (takanome-dev) [inline src/user/services/user/user.service.spec.ts:227]: cc @brandonroberts @AlexHladin 

I tried `expect(dbUserRepositoryMock.save).toHaveBeenCalled()` but it didn't work because it's not the `userRepository.save()` that has been called but `newUser.save()` which is an instance of that class

File: src/user/services/user/user.service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+  describe("findOneById", () => {
+    const user = { id: faker.number.int(), email: faker.internet.email() };
+    const createQueryBuilderMock = {
+      addSelect: jest.fn().mockReturnThis(),
+      where: jest.fn().mockReturnThis(),
+      setParameters: jest.fn().mockReturnThis(),
+      getOne: jest.fn().mockResolvedValue(user),
+    };
+
+    /**
```
Comment (takanome-dev) [inline src/user/services/user/user.service.spec.ts:92]: I removed the test suite when `includeEmail` is false because it always includes the email. IMO it's because I mocked the response. Maybe this should be tackled when doing e2e

File: src/user/services/user/user.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+     * @author @takanome-dev
+     * @todo test with a real db when doing e2e if `findOneById`
+     * with includeEmail=false is returning the email or not.
+     * Cannot be tested with unit test (IMO) since the result is mocked.
+     */
```
Comment (takanome-dev) [inline src/user/services/user/user.service.spec.ts:97]: I removed the test suite when `includeEmail` is false because it always includes the email. IMO it's because I mocked the response. Maybe this should be tackled when doing e2e

File: src/user/services/user/user.service.spec.ts (hunk 1 above)
Comment (AlexHladin) [reply src/user/services/user/user.service.spec.ts:227]: @takanome-dev you can do a simple thing:
```typescript
const saveMock = jest.fn()
dbUserRepositoryMock.create.mockReturnValue(createMock)
...
expect(saveMock).toHaveBeenCalled()
```

**Comment by takanome-dev** [comment]
@brandonroberts I didn't add tests for the controller because it's just the same mocks and tests as the service (just duplicates). It would be better to test them when doing e2e.

That's just my opinion, do let me know if you think I should do it.

**Comment by brandonroberts** [comment]
@takanome-dev seems reasonable to me

File: src/user/services/user/user.service.ts [hunk 4]
```diff
@@ -3,27 +3,25 @@ import { Repository, SelectQueryBuilder } from "typeorm";
...
+import { PageMetaDto } from "../../../common/dtos/page-meta.dto";
 
 @Injectable()
 export class UserService {
   constructor(
     @InjectRepository(DbUser, "ApiConnection")
     private userRepository: Repository<DbUser>,
     @InjectRepository(DbUserHighlightReaction, "ApiConnection")
-    private userHighlightReactionRepository: Repository<DbUserHighlightReaction>,
-    private pagerService: PagerService
```
Comment (jpmcb) [inline src/user/services/user/user.service.ts:26]: I'm guessing the `pagerService` was unused in this service?

**Comment by jpmcb** [comment]
Is there a related tracking issue or ticket for this? If not, let's make one so we can track testing efforts across the API.

> Note that user.service has been moved to a new folder services/user for a better structure. The other services will be moved in future prs.

I'd like to understand this more: what benefit do we get from moving the services under a `services/` directory? Better structure in the code? Is this a nest js convention?

File: src/user/services/user/user.service.ts (hunk 4 above)
Comment (takanome-dev) [reply src/user/services/user/user.service.ts:26]: Yeah, but it should be used to remove duplicates. We should open an issue about cleaning up the usage of `new PageMetaDto()` and `new PageDto()` in services.

**Comment by takanome-dev** [comment]
> Is there a related tracking issue or ticket for this? If not, let's make one so we can track testing efforts across the API.
> 
Yeah -> https://github.com/open-sauced/api/issues/215

> I'd like to understand this more: what benefit do we get from moving the services under a `services/` directory? Better structure in the code? Is this a nest js convention?
>
Just for better structure. There are 4 other services that need to be tested. Dumping all the test files in the user module may be messy IMO.

**Comment by brandonroberts** [comment]
I don't think we need to nest the folders too much. There's already a `user` folder, no need to have a `services/user` folder with one service inside there.

If we have `user/services` and all the services are in there that should be good enough. We don't want to create a complex folder structure that's too granular.

**Comment by takanome-dev** [comment]
> If we have `user/services` and all the services are in there that should be good enough. We don't want to create a complex folder structure that's too granular.

@brandonroberts done :+1:

File: src/user/services/user.service.ts [hunk 5]
```diff
@@ -176,7 +174,7 @@ export class UserService {
       return user;
     } catch (e) {
       // create new user
-      const newUser = this.userRepository.create({
+      const newUser = await this.userRepository.save({
```
Comment (brandonroberts) [inline src/user/services/user.service.ts:177]: Revert this

**Review by brandonroberts — APPROVED** [review]
LGTM. Revert the user service and open a separate PR unless it impacts the tests

**Comment by takanome-dev** [comment]
> LGTM. Revert the user service and open a separate PR unless it impacts the tests

Yeah it does, as mentioned [here](https://github.com/open-sauced/api/pull/219#discussion_r1270478390) it was failing the test and I couldn't figure out a way to mock that instance.

**Comment by brandonroberts** [comment]
@takanome-dev got it. Going forward, we shouldn't have to modify existing working code to make tests pass

**Comment by takanome-dev** [comment]
> @takanome-dev got it. Going forward, we shouldn't have to modify existing working code to make tests pass

Sure, will keep that in mind. Thanks :+1:

**Review by OgDev-01 — APPROVED** [review]
